flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/TimeBoundedStreamJoinOperatorTest.java

```diff
@@ -0,0 +1,590 @@
+			assertOutput(expected, testHarness.getOutput());
+			ensureNoLateData(testHarness.getOutput());
+		}
+	}
+
+	@Test
+	public void testNegativeExclusiveAndNegativeExlusive() throws Exception {
+		long lowerBound = -3;
+		boolean lowerBoundInclusive = false;
+
+		long upperBound = -1;
+		boolean upperBoundInclusive = false;
+
+		try (KeyedTwoInputStreamOperatorTestHarness<String, TestElem, TestElem, Tuple2<TestElem, TestElem>> testHarness
+				 = createTestHarness(lowerBound, lowerBoundInclusive, upperBound, upperBoundInclusive)) {
+
+			testHarness.setup();
+			testHarness.open();
+			prepareTestHarness(testHarness);
+
+			List<StreamRecord<Tuple2<TestElem, TestElem>>> expectedOutput = Lists.newArrayList(
+				streamRecordOf(3, 1),
+				streamRecordOf(4, 2)
+			);
+
+			ConcurrentLinkedQueue<Object> output = testHarness.getOutput();
+
+			assertOutput(expectedOutput, testHarness.getOutput());
+			ensureNoLateData(output);
+		}
+	}
+
+	@Test
+	public void testNegativeExclusiveAndPositiveExlusive() throws Exception {
+		long lowerBound = -1;
+		boolean lowerBoundInclusive = false;
+
+		long upperBound = 1;
+		boolean upperBoundInclusive = false;
+
+		try (KeyedTwoInputStreamOperatorTestHarness<String, TestElem, TestElem, Tuple2<TestElem, TestElem>> testHarness
+				 = createTestHarness(lowerBound, lowerBoundInclusive, upperBound, upperBoundInclusive)) {
+
+			testHarness.setup();
+			testHarness.open();
+			prepareTestHarness(testHarness);
+
+			List<StreamRecord<Tuple2<TestElem, TestElem>>> expectedOutput = Lists.newArrayList(
+				streamRecordOf(1, 1),
+				streamRecordOf(2, 2),
+				streamRecordOf(3, 3),
+				streamRecordOf(4, 4)
+			);
+
+			ConcurrentLinkedQueue<Object> output = testHarness.getOutput();
+
+			assertOutput(expectedOutput, testHarness.getOutput());
+			ensureNoLateData(output);
+		}
+	}
+
+	@Test
+	public void testPositiveExclusiveAndPositiveExlusive() throws Exception {
+		long lowerBound = 1;
+		boolean lowerBoundInclusive = false;
+
+		long upperBound = 3;
+		boolean upperBoundInclusive = false;
+
+		try (KeyedTwoInputStreamOperatorTestHarness<String, TestElem, TestElem, Tuple2<TestElem, TestElem>> testHarness
+				 = createTestHarness(lowerBound, lowerBoundInclusive, upperBound, upperBoundInclusive)) {
+
+			testHarness.setup();
+			testHarness.open();
+			prepareTestHarness(testHarness);
+
+			List<StreamRecord<Tuple2<TestElem, TestElem>>> expectedOutput = Lists.newArrayList(
+				streamRecordOf(1, 3),
+				streamRecordOf(2, 4)
+			);
+
+			ConcurrentLinkedQueue<Object> output = testHarness.getOutput();
+
+			assertOutput(expectedOutput, testHarness.getOutput());
+			ensureNoLateData(output);
+
+		}
+	}
+
+	@Test
+	public void stateGetsCleanedWhenNotNeeded() throws Exception {
+
+		long lowerBound = 1;
+		boolean lowerBoundInclusive = true;
+
+		long upperBound = 2;
+		boolean upperBoundInclusive = true;
+
+		TimeBoundedStreamJoinOperator<TestElem, TestElem> operator = new TimeBoundedStreamJoinOperator<>(
+			lowerBound,
```

Comment:
Here, and wherever we instantiate an operator, we should now add the serializers for the input elements. This can be done with the `TypeInformation.of(new TypeHint<TestElem>() {}).createSerializer(new ExecutionConfig())`.